Dev step3

wuctlby opened the pull request:
Hi @stefanopolitano and @Marcellocosti , this PR is about the step3, including [src/cut_variation.py](https://github.com/flowHF/hf-vn/commit/a6bc8db0f80fc46c0c7cf046a89d2f6f3e39f637), [src/data_driven_fraction.py](https://github.com/flowHF/hf-vn/commit/d37252c7d003df988f1aacc6c02fcb42f0521647), [src/get_v2_vs_frac.py](https://github.com/flowHF/hf-vn/commit/92636937bcf46ec2d9d645be1aed2e3961126a90) and [utils/frac_utils.py](https://github.com/flowHF/hf-vn/commit/e06248480784379ae915856420131980e7458908).

Basically, they are all `new` scripts to adapt the new framework. Unless the systematic fraction haven't been tested yet, other should the fine. The uncorrelated method is not surppoted any more in the main workflow. 

I also want to move all the load methods used to load inside files into a new `utils` script. If you have a better way of doing this, please let me know.

It seems that all the commits were submitted in this PR. Hope that's OK. Maybe the tests could be done one PR at a time. Please let me know if there are any problems. Thank you in advance for the reveiw!

wuctlby committed:
Add full workflow on run_analysis.py

wuctlby committed:
updates on yml file: full operations and using list for axis to keep

wuctlby committed:
updates on sparse_dicts.py: add axes dict for D0's reflection; using the list for axis to keep

wuctlby committed:
update on make_cutsets_cfgs.py: in case the pt bins is less the number of bkg cuts

wuctlby committed:
update on src/pre_process.py

wuctlby committed:
update on src/proj_thn.py

wuctlby committed:
updates on src/compute_efficiencies.py

wuctlby committed:
updates on src/get_vn_vs_mass.py

wuctlby committed:
updates on utils/utils.py

wuctlby committed:
new: src/cut_variation.py

wuctlby committed:
new: src/data_driven_fraction.py

wuctlby committed:
new: src/get_v2_vs_frac.py

wuctlby committed:
new: utils/frac_utils.py

wuctlby commented:
Not only the systematic fraction but also the output path needs to be tested and corrected to the desired path. Currently, it uses the wrong path, but it won't affect the usage.

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+                        default="input", help="input path to the raw yields")
+    parser.add_argument('infilePathFrac', metavar='text',
+                        default="input", help="input path to the data driven fractions")
+    parser.add_argument('--correlated', '-corr', action='store_true',
```
I see that you use the `--correlated` flag to count the cutsets and obtain `nSets`, but this in principle should be possible by counting the raw yields or fraction files. What do you think?

wuctlby replied:
Yes, you reminded me that we can't count the `nSets` from the number of raw yields or fractions. We should also correct it in `cut_variation.py`. Because, for some pt bins, we will duplicate the process of some cuts, we have to drop them in the steps which are after the cut variation. Or have we already fixed this?

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+            sig = cfg['cut_variation']['uncorr_bdt_cut']['sig']
+            CutSets = [len(sig[i]) - 1 for i in range(nPtBins)]
+
+    if len(fracFiles) != len(rawYieldFiles):
```
We can move this check directly in `main_v2_vs_frac` before even entering the function, what do you think?

wuctlby replied:
Yes, you are right

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+    outFileName = os.path.join(parentDir, dirName, fileName)
+    outputdir = os.path.dirname(outFileName)
+    os.makedirs(outputdir, exist_ok=True)
+    outFile = TFile(outFileName, "recreate")
```
Can't this just be replaced by
```suggestion
    outputdir = os.path.dirname(fracFiles[0]).replace('frac', 'v2VsFrac')
    os.makedirs(outputdir, exist_ok=True)
    outFile = TFile(f"{outputdir}/v2VsFrac.root", "recreate")
```

wuctlby replied:
Here, I would like to avoid replacing it directly. I think replacing 'frac' or other 'key' words directly is not safe. What if they exist in the path. But you are right, some cases we can just directly use the name, since there is no `suffix` anymore.

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+        particleName = cfg['Dmeson']
+        particleTit, _, decay, _, _, _ = get_particle_info(particleName)
+        import numpy as np
+        if correlated:
+            sig = cfg['cut_variation']['corr_bdt_cut']['sig']
+            CutSets = [len(list(np.arange(sig['min'][i], sig['max'][i], sig['step'][i]))) - 1 for i in range(nPtBins)]
+        else:
+            sig = cfg['cut_variation']['uncorr_bdt_cut']['sig']
+            CutSets = [len(sig[i]) - 1 for i in range(nPtBins)]
```
Maybe we can avoid this and initialize `nSets` from the number of files which we obtain for fractions and yields, e.g. by calling `len(rawYieldFiles)`?

wuctlby replied:
I replied to it in detail in the comment above. Generally, we will have duplicate cuts, so we can't count them from the number of files.

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+
+    cFrac, ptStrings, chi2Strings = [], [], []
+    
+    nPtBins = len(ptmins)
```
I think that this line is repeated when the `config_flow` file is read

wuctlby replied:
Yes, you are right

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+        SetObjectStyle(gFracVsV2[-1], linecolor=kAzure+4, linewidth=2, markerstyle=kFullCircle, markersize=1, markercolor=kAzure+4)
+
+        print(f"Processing pt bin {iPt+1}/{nPtBins}: {ptMin:.2f} < pT < {ptMax:.2f} GeV/c, nSets: {nSets}")
+        avrV2XErrL.append(Double_t(sum(gV2[i].GetErrorXlow(iPt) for i in range(nSets)) / nSets))
+        avrV2XErrH.append(Double_t(sum(gV2[i].GetErrorXhigh(iPt) for i in range(nSets)) / nSets))
```
I see that `avrV2XErrL` and `avrV2XErrH` are never used, right? I think that we can remove them both here and when they are initialized as empty lists

wuctlby replied:
Yes, this is used to mark the error range of pT. Since now we always use the pt bin center, we can remove it. And when we have the mean pt shift, we add them back.

Marcellocosti commented on src/data_driven_fraction.py:
```diff
@@ -0,0 +1,236 @@
+    cutVarFracFile   = TFile.Open(cutVarFracFile)
+
+    hCorrYieldPrompt = cutVarFracFile.Get('hCorrYieldPrompt')
+    hCorrYieldPrompt.SetDirectory(0)
```
Can you try to comment all the lines with the command `SetDirectory(0)` and set `ROOT.TH1.AddDirectory(False)` globally where we perform the imports via the line? If it works we can then remove all these `SetDirectory(0)` lines here and there!

wuctlby replied:
Nice point!, I will use the way you suggested.

Marcellocosti commented on src/data_driven_fraction.py:
```diff
@@ -0,0 +1,236 @@
+from frac_utils import GetPromptFDFractionCutSet  # pylint: disable=import-error,no-name-in-module # pyignore # type: ignore
+from ROOT import TFile, TCanvas, TLegend, gROOT, kRed, kBlue  # pylint: disable=import-error,no-name-in-module # pyignore # type: ignore
+from StyleFormatter import SetGlobalStyle # pylint: disable=import-error,no-name-in-module # pyignore # type: ignore
+
```
```suggestion
ROOT.TH1.AddDirectory(False)
```

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+                                hV2VsFrac[-1].GetBinError(hV2VsFrac[-1].GetNbinsX()))
+        
+        # get the v2 value at the FD fraction = 0, and it is the first bin?
+        hV2VsPtPrompt.SetBinContent(iPt + 1, 
+                                    hV2VsFrac[-1].GetBinContent(hV2VsFrac[-1].GetBin(1)))
+        hV2VsPtPrompt.SetBinError(iPt + 1,
+                                    hV2VsFrac[-1].GetBinError(hV2VsFrac[-1].GetBin(1)))
```
```suggestion
        hV2VsPtPrompt.SetBinContent(iPt + 1, 
                                    hV2VsFrac[-1].GetBinContent(1))
        hV2VsPtPrompt.SetBinError(iPt + 1,
                                    hV2VsFrac[-1].GetBinError(1))
```

wuctlby replied:
Here, I will first implement what you suggested. I checked it, and they are the same. 
```
getbin 1: 0.0317836114624695, bin: 1
findbin 0: 0.0317836114624695, bin: 1
getbincontent 1: 0.0317836114624695
function: 0.031818099614257474
```
Because at the beginning, I inherited it from what we did in run2. And the thing that is not clear to me is why we don't just use the value at x=0 in the fit function. So I didn't change it. Do you have any suggestions?@stefanopolitano ?

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+
+    gROOT.SetBatch(True)
+    
+    #!!!!!!!! totall drop the uncorrelated cutsets change /home/wuct/ALICE/local/dev/hf-vn/src/cut_variation.py
```
```suggestion
```

wuctlby replied:
Sorry, I forgot to delete this comment. It was just used to remind me to change the script accordingly.

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+        ptmaxs = cfg['ptbins'][1:]
+        nPtBins = len(ptmins)
+        particleName = cfg['Dmeson']
+        particleTit, _, decay, _, _, _ = get_particle_info(particleName)
```
```suggestion
        particleTit, _, decay, _, _, _ = get_particle_info(cfg['Dmeson'])
```

so we can also avoid initializing `particleName = None` above

Marcellocosti commented on src/get_v2_vs_frac.py:
```diff
@@ -0,0 +1,298 @@
+        raise ValueError(f'Number of eff and frac files do not match: {len(fracFiles)} != {len(rawYieldFiles)}')
+
+    hV2, gV2, hFracFD, hFracPrompt = [], [], [], []
+    avrV2XErrL, avrV2XErrH = [], []
```
It seems to me that we never use such lists

Marcellocosti commented on src/data_driven_fraction.py:
```diff
@@ -0,0 +1,236 @@
+    dirName   = os.path.dirname(cutVarFile).split('/')[-1].replace('cutVar', 'frac') # cutVar ==> frac
+    outputDir = os.path.join(parentDir, dirName)                                       # ./cutvar_suffix/frac
+    logger(f'Output directory: {outputDir}', level='INFO')
+    os.makedirs(os.path.dirname(outputDir), exist_ok=True)
```
Can you try

```suggestion
    os.makedirs(os.path.dirname(cutVarFile).replace('cutVar', 'frac'), exist_ok=True)
```

wuctlby replied:
Just llike what I commented on the `frac` case, for me, it seems it's not a safe way.

Marcellocosti commented on src/data_driven_fraction.py:
```diff
@@ -0,0 +1,236 @@
+                        dirNameSyst = os.path.dirname(cutVarSystFile).split('/')[-1].replace('cutVar', 'frac') # cutVar ==> frac
+                        outputDirSyst = os.path.join(os.path.dirname(root), dirNameSyst) # ./cutvar_suffix/syst_frac/sysOpt/frac
+                        logger(f'Output directory for systematic fraction: {outputDirSyst}', level='INFO')
+                        os.makedirs(os.path.dirname(outputDirSyst), exist_ok=True)
```
If I understood correctly, @stefanopolitano was working on the code of the systematic relative to the fraction, so maybe it is more convenient not to merge this piece of code for now and wait for the PR of Stefano with the full implementation. What do you guys think?

By the way, I also think that since this script takes `cutVarFile` and `effPath`, maybe the cleanest solution is not to add separate code for systematics here code but just launch this script pointing it to the folders with all cases (and so remove also the `syst_frac` from the arguments).

wuctlby replied:
As you wish, I will clean all the parts of the systematic fraction. I also think it would be better to have a separate script to trigger the systematic fractions.

Marcellocosti commented on src/cut_variation.py:
```diff
@@ -0,0 +1,419 @@
+import numpy as np # type: ignore
+from itertools import product
+import ROOT
+from ROOT import TFile, TCanvas, TLegend, TLatex, gROOT, gStyle
+from ROOT import TFile, TH1F, TH2F, TCanvas, TLegend, TLatex, gRandom  # pylint: disable=import-error,no-name-in-module
```
I see that some imports are duplicated

Marcellocosti commented on src/cut_variation.py:
```diff
@@ -0,0 +1,419 @@
+                listEffPromptUnc = listEffPromptUnc[2:]
+                listEffFD        = listEffFD[2:]
+                listEffFDUnc     = listEffFDUnc[2:]
+            # Remove the tightest 3 cuts if 'minus3high' or 'minus' is in systOpt
```
I think that, if @stefanopolitano is taking care of the implementation of this systematics, all this block can be removed from the current PR so the relevant changes are better highlighted

Marcellocosti commented on src/cut_variation.py:
```diff
@@ -0,0 +1,419 @@
+        ptmins = config['ptbins'][:-1]
+        ptmaxs = config['ptbins'][1:]
+
+    if os.path.exists(inputPathEff):
+        effFiles = [f'{os.path.join(inputPathEff, file)}'
+                    for file in os.listdir(inputPathEff) if file.startswith('eff_') and file.endswith('.root')]
+    else:
+        logger(f'No eff folder found in {inputPathEff}', level='ERROR')
+        raise ValueError(f'No eff folder found in {inputPathEff}')
+
+    if os.path.exists(inputPathRy):
+        rawYieldFiles = [f'{os.path.join(inputPathRy, file)}'
+                         for file in os.listdir(inputPathRy) if file.startswith('raw_yields_') and file.endswith('.root')]
+    else:
+        logger(f'No vn folder found in {inputPathRy}', level='ERROR')
+        raise ValueError(f'No vn folder found in {inputPathRy}')
```
```suggestion
    if not os.path.exists(inputPathEff):
            logger(f'No eff folder found in {inputPathEff}', level='ERROR')
            raise ValueError(f'No eff folder found in {inputPathEff}')
    if not os.path.exists(inputPathRy):
            logger(f'No vn folder found in {inputPathRy}', level='ERROR')
            raise ValueError(f'No vn folder found in {inputPathRy}')
        
    effFiles = [f'{os.path.join(inputPathEff, file)}'
                    for file in os.listdir(inputPathEff) if file.startswith('eff_') and file.endswith('.root')]
    rawYieldFiles = [f'{os.path.join(inputPathRy, file)}'
                         for file in os.listdir(inputPathRy) if file.startswith('raw_yields_') and file.endswith('.root')]
```

I suggest this version as it will fail immediately and we diminish the nested loops

wuctlby replied:
True, and I will move all of them into a separate script in `utils`.

Marcellocosti commented on src/cut_variation.py:
```diff
@@ -0,0 +1,419 @@
+    latInfo.SetTextColor(1)
+
+    minimise_chi2(config, ptmins, ptmaxs, hRawYields, hEffPrompt, hEffFD, inputPathEff, correlated=correlated)
+    if config['operations']['do_syst_frac'] and correlated:
+        for syst in config['minimisation']['systematics']:
+            logger(f'Running cut variation for systematics: {syst}')
+            minimise_chi2(config, ptmins, ptmaxs, hRawYields, hEffPrompt, hEffFD, inputPathEff, correlated=correlated, systOpt=syst)
```
Maybe better to have a separate PR for the fraction systematics, to be discussed also with @stefanopolitano

Marcellocosti commented on src/cut_variation.py:
```diff
@@ -0,0 +1,419 @@
+                        default="input", help="input path to the raw yields")
+    parser.add_argument("infilePathEff", metavar="text",
+                        default="input", help="input path to the efficiencies")
+    parser.add_argument("--correlated", "-c", action="store_true",
+                        help="Produce cut variation files for correlated cuts")
```
```suggestion
```

I see that the `correlated` argument is never used in the functions, and that is very nice as we only need the paths of the ry and eff folders. Then we can remove such flag in my opinion

wuctlby replied:
Yes, since the uncorrelated cuts method is not supported anymore, it should be removed. But I kept it for the systematic fraction. Now Stefano will take care of it, I will also remove it.

Marcellocosti commented:
Hi @wuctlby, thanks a lot for the excellent work! I left some comments, mainly minor things regarding some variables/flags which we do not need and that can be removed in my opinion. Another point is the systematics on the fraction: I think @stefanopolitano is preparing something, so if this is confirmed I suggest that we remove such changes from this PR and have a dedicated one for that part of the code so that it is easier to look at the new code. Thanks again!

wuctlby commented:
> Hi @wuctlby, thanks a lot for the excellent work! I left some comments, mainly minor things regarding some variables/flags which we do not need and that can be removed in my opinion. Another point is the systematics on the fraction: I think @stefanopolitano is preparing something, so if this is confirmed I suggest that we remove such changes from this PR and have a dedicated one for that part of the code so that it is easier to look at the new code. Thanks again!

Hi @Marcellocosti, thanks a lot for the review! Since there are too many comments, I will not summarise them here. I will directly fix them. And some of them we may have different opinions. Please let me know if you have further suggestions for those comments! Thanks again!

wuctlby committed:
Implement the comments and add the readme to explain the output

wuctlby committed:
configuration file

wuctlby committed:
remove `syst_frac`

wuctlby committed:
Update on readme

Marcellocosti commented on run_analysis.py:
```diff
@@ -78,81 +83,211 @@ def run_fit(i):
 		results_fit = list(executor.map(run_fit, range(mCutSets)))
 
-def run_correlated_cut_variation(config, operations, nworkers, outdir):
+def cut_variaion(flow_config, outdir, correlated, combined=False, operations=None, batch=False):
```
```suggestion
def cut_variation(flow_config, outdir, correlated, combined=False, operations=None, batch=False):
```
Just a small typo

Marcellocosti commented on src/compute_efficiencies.py:
```diff
@@ -132,8 +138,10 @@ def compute_eff(config, inputFile, batch=False):
+    outFileName = os.path.join(os.path.dirname(os.path.dirname(inputFile)),
+                               'eff',
+                               os.path.basename(inputFile).replace('proj', 'eff'))
+    os.makedirs(os.path.dirname(outFileName), exist_ok=True)
```
Correct me if I am wrong, but I think that the `os.makedirs` in the scripts in the `src/` folder are all redundant as we apply `check_dir` in the `run_analysis.py` script. Can you try if if works when removing them?

wuctlby replied:
Yes, here the function of `os.makdirs` and `check_dir` is different. As you saw, the former one is used in the script to make sure the output path exists when running this script. The latter one is used to check if there is already output, if so, we remove and recreate it. Even though in a sense they are redundant, but I would like to keep them. What do you think?

Marcellocosti replied:
It seems to me that `check_dir` checks whether the directory exists or not. If it doesn't exist it creates the folder, if it exists it clears and recreates the folder, right? If it is like this I think this is fully a duplicate, right? But I think that we can merge the PR and make this small adjustments in other PRs.

wuctlby replied:
Yes, the function is like what you said. And if we remove the `os.makedirs` in the script, `run_analysis.py` should also work. But the story is like this: 
First, sometimes we have to test a set of configs several times in the same output path. In case we should avoid using the previous results, like cutset configs. Then we have this function `check_dir` to recreate the directory no matter the directory is there or not.
Second, sometimes, we may want to test or run the separate script, at least for me. So it's necessary to have this kind of `os.makedirs` in the script to make sure that when we run the script separately, it also works.
So, last, `check_dir` is more like a help function in `run_analysis.py` to run the whole workflow correctly.

Marcellocosti commented on src/cut_variation.py:
```diff
@@ -0,0 +1,337 @@
+    for iRow, row in enumerate(hCovCorrYields):
+        for iCol, hCov in enumerate(row):
+            SetObjectStyle(hCov, linecolor=kBlack)
+            if iRow == 0:
+                rowName = '#it{N}_{prompt}'
+            else:
+                rowName = '#it{N}_{non-prompt}'
+            if iCol == 0:
+                colName = '#it{N}_{prompt}'
+            else:
+                colName = '#it{N}_{non-prompt}'
```
```suggestion
                rowName = '#it{N}_{prompt}' if iRow == 0 else '#it{N}_{non-prompt}'
                colName  = '#it{N}_{prompt}' if iCol == 0   else '#it{N}_{non-prompt}'
```

Marcellocosti commented on src/cut_variation.py:
```diff
@@ -0,0 +1,337 @@
+                        default="input", help="input path to the raw yields")
+    parser.add_argument("infilePathEff", metavar="text",
+                        default="input", help="input path to the efficiencies")
+    parser.add_argument("--batch", "-b", action="store_true",
+                        help="run in batch mode")
```
Question for you guys: do we really need this `--batch` flag in the scripts? Personally I have never used it and just let the program go. If no one of us ever uses it, we can remove it, but of course if you use it we keep it :)

wuctlby replied:
You are right. Essentially, we do not need the `--batch` option when using `run_analysis.py`. However, sometimes, if we need to test something separately, we may need to use this to examine the results. So I will set this option to `true` in the `run_analysis.py`, but we keep it in the arguments of the script. What do you think?

Marcellocosti replied:
Nevermind, we can leave it like this if you use it sometimes!

Marcellocosti commented:
From my side, green light. Thanks a lot @wuctlby for addressing the comments!

wuctlby committed:
using `logger` and set batch as true for the default case

wuctlby commented:
> From my side, green light. Thanks a lot @wuctlby for addressing the comments!

Hi @Marcellocosti , thanks a lot! The comments you mentioned have been addressed in the new commit.

stefanopolitano commented:
Hi @wuctlby and @Marcellocosti! Thanks for the development. I am merging the pr to test it locally
